### PR TITLE
Improve temp directory handling in test suite

### PR DESCRIPTION
**Pull request by pitrou**
This should allow to run several CI builds at once without any conflicts.

### CODE REVIEW

File: numba/tests/support.py [hunk 1]
```diff
@@ -409,27 +411,63 @@ def tweak_code(func, codestring=None, consts=None):
...
+    _trashcan_dir = os.path.join(tempfile.gettempdir(), _trashcan_dir)
+else:
+    # Mix the UID into the directory name to allow different users to
+    # run the test suite without permission errors (issue #1586)
+    _trashcan_dir = os.path.join(tempfile.gettempdir(),
+                                 "%s.%s" % (_trashcan_dir, os.getuid()))
+
+# Stale temporary directories are deleted after they are older than this value.
+# The test suite probably won't ever take longer than this...
+_trashcan_timeout = 24 * 3600
```
Comment (stefanseefeld) [inline numba/tests/support.py]: I gather this is 'one day'. It would be good to spell that out, if only in comment.

File: numba/tests/support.py [hunk 2]
```diff
@@ -409,27 +411,63 @@ def tweak_code(func, codestring=None, consts=None):
...
+    Create a temporary directory with the given *prefix* that will survive
+    at least as long as this process invocation.  The temporary directory
+    will be eventually deleted when it becomes stale enough.
+
+    This is necessary because a DLL file can't be deleted while in use
+    under Windows.
+
+    An interesting side-effect is to be able to inspect the test files
+    shortly after a test suite run.
+    """
```
Comment (stefanseefeld) [inline numba/tests/support.py]: Thanks for that rationale. I would still prefer if by default test artifacts (such as temporary directories) were cleaned up after each run, at least on sane platforms that allow that. That being said, I agree to the usefulness to be able to keep such data around for forensic analysis, but I would prefer that to be done through a command-line option (--keep-temporaries ?).

Comment (pitrou) [inline numba/tests/support.py]: IMO that would only complicate the code for little gain (sparing a couple MB accross runs perhaps?).

File: numba/tests/support.py (hunk 1 above)
Comment (pitrou) [inline numba/tests/support.py]: Ok, done.

**Pull request merged**